www/js/app.js

```diff
@@ -1051,16 +1074,30 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
                     $("#searchingArticles").show();
                 };
             };
+
+            if(! isDirEntryExpectedToBeDisplayed(dirEntry)){
+                console.debug(dirEntry, " is not expected to be displayed. Early retrun in readArticle().")
+                return;
+            } 
+
             // We put the ZIM filename as a prefix in the URL, so that browser caches are separate for each ZIM file
             iframeArticleContent.src = "../" + selectedArchive._file._files[0].name + "/" + dirEntry.namespace + "/" + encodedUrl;
         } else {
             // In jQuery mode, we read the article content in the backend and manually insert it in the iframe
             if (dirEntry.isRedirect()) {
-                selectedArchive.resolveRedirect(dirEntry, readArticle);
+                // Update expectedArticleURLToBeDisplayed before redirect then call readArticle.
+                selectedArchive.resolveRedirect(dirEntry, function (resolvedDirEntry) {
+                    expectedArticleURLToBeDisplayed = resolvedDirEntry.namespace + "/" + resolvedDirEntry.url;
+                    readArticle(resolvedDirEntry)
+                });
             } else {
                 // Line below was inserted to prevent the spinner being hidden, possibly by an async function, when pressing the Random button in quick succession
                 // TODO: Investigate whether it is really an async issue or whether there is a rogue .hide() statement in the chain
                 $("#searchingArticles").show();
+                if(! isDirEntryExpectedToBeDisplayed(dirEntry)){
```

Comment:
It's still too early to do that.
Let me re-explain our goal : we need to assign the variable as soon as possible (when we know the URL that is requested), and we need to check if this variable did not change right before displaying the article content. So that the main asynchronous part (reading the content in the ZIM file) is between these 2 steps.
Here, `selectedArchive.readUtf8File` is the asynchronous part, so you need to move that at the beginning of `displayArticleContentInIframe`

Reply:
Let me try to rephrase this problem to see if I truly understand the design here. 

Consider the following situation with without refactoring the too early early return in the code. 
Assumes the user clicked on article **1**.  The user does not click on article **2** until function `displayArticleContentInIframe()` with dirEntry for article **1** as it's input into the asyn queue.  With the too early early return,  article **1** is still loaded before article **2** which is not expected in this rare condition under jQuery mode.  This can be fixed by placing the early return right at the begining of `displayArticleContentInIframe`. 

As a side note, there is no such problem in serviceworkder mode since there is no asyn loading.